### Import Necessary libraries

In [1]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

#### Import Dataset

In [2]:
glass_categorizing = pd.read_csv('glass.csv')
glass_categorizing

RI     Na    Mg    Al     Si     K    Ca    Ba   Fe  Type
0    1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.00  0.0     1
1    1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.00  0.0     1
2    1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.00  0.0     1
3    1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.00  0.0     1
4    1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.00  0.0     1
..       ...    ...   ...   ...    ...   ...   ...   ...  ...   ...
209  1.51623  14.14  0.00  2.88  72.61  0.08  9.18  1.06  0.0     7
210  1.51685  14.92  0.00  1.99  73.06  0.00  8.40  1.59  0.0     7
211  1.52065  14.36  0.00  2.02  73.42  0.00  8.44  1.64  0.0     7
212  1.51651  14.38  0.00  1.94  73.61  0.00  8.48  1.57  0.0     7
213  1.51711  14.23  0.00  2.08  73.36  0.00  8.62  1.67  0.0     7

[214 rows x 10 columns]

### Data Understanding

In [3]:
glass_categorizing.shape

(214, 10)

In [4]:
glass_categorizing.isna().sum()

RI      0
Na      0
Mg      0
Al      0
Si      0
K       0
Ca      0
Ba      0
Fe      0
Type    0
dtype: int64

In [7]:
glass_categorizing.dtypes

RI      float64
Na      float64
Mg      float64
Al      float64
Si      float64
K       float64
Ca      float64
Ba      float64
Fe      float64
Type      int64
dtype: object

In [8]:
glass_categorizing.dtypes

RI      float64
Na      float64
Mg      float64
Al      float64
Si      float64
K       float64
Ca      float64
Ba      float64
Fe      float64
Type      int64
dtype: object

### Model Building

In [9]:
X = glass_categorizing.drop(labels=['Type'],axis = 1)
y = glass_categorizing[['Type']]

In [10]:
X.shape,y.shape

((214, 9), (214, 1))

In [11]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.15,random_state=12)

In [12]:
X_train.shape,y_train.shape

((181, 9), (181, 1))

In [13]:
X_test.shape,y_test.shape

((33, 9), (33, 1))

##  Model Training | Testing | Evaluation

before standardizing

In [17]:
from sklearn.neighbors import KNeighborsClassifier
knn_classifier = KNeighborsClassifier(n_neighbors=3)
knn_classifier.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=3)

In [18]:
knn_classifier = KNeighborsClassifier(n_neighbors=3)
knn_classifier.fit(X_train,y_train)

#Training data
y_pred_train = knn_classifier.predict(X_train)
print('Accuracy Score:',round(accuracy_score(y_train,y_pred_train),4))

Accuracy Score: 0.8398


##  Model Training | Testing | Evaluation

In [19]:
import warnings
warnings.filterwarnings('ignore')

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

In [20]:
from sklearn.preprocessing import StandardScaler
std_scaler = StandardScaler()
std_scaler.fit(glass_categorizing.drop('Type',axis=1))
scaled_features = std_scaler.transform(glass_categorizing.drop('Type',axis=1))
X_scaled = pd.DataFrame(scaled_features,columns=glass_categorizing.columns[:-1])

In [21]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_scaled,y,test_size=0.15,random_state=12,stratify=y)

In [22]:
knn_classifier = KNeighborsClassifier(n_neighbors=2)
knn_classifier.fit(X_train,y_train)

#Training data
y_pred_train = knn_classifier.predict(X_train)
print('Accuracy Score:',round(accuracy_score(y_train,y_pred_train),4))

Accuracy Score: 0.8287


In [23]:
knn_classifier = KNeighborsClassifier(n_neighbors=3)
knn_classifier.fit(X_train,y_train)

#Training data
y_pred_train = knn_classifier.predict(X_train)
print('Accuracy Score:',round(accuracy_score(y_train,y_pred_train),4))

Accuracy Score: 0.8287


In [24]:
knn_classifier = KNeighborsClassifier(n_neighbors=4)
knn_classifier.fit(X_train,y_train)

#Training data
y_pred_train = knn_classifier.predict(X_train)
print('Accuracy Score:',round(accuracy_score(y_train,y_pred_train),4))

Accuracy Score: 0.8011


**One neighbor is optimal for the data with a accuracy of 0.8287**